### DISCUSSION TITLE
Help with row level security

### DISCUSSION THREAD

**Discussion by tobbbe** [supabase/supabase#3695] [Questions] [answered]
Hello!
I have 3 tables:

**videos**
id
name
project_id

**projects**
id
name

**project_members**
project_id
user_id

I want a rule that allows a user to only access videos for projects the user belongs to.

I've tried this for example, doesnt work:
```
(
  EXISTS (
    SELECT
      1
    FROM
      project_members
    WHERE
      (
        (project_members.user_id = uid())
        AND (project_members.project_id = videos.project_id)
      )
  )
)
```

I also want a user to be able to get all projects she belongs to.

Please help! I'm ripping my hair of :(

**Comment by soedirgo** [comment, marked as answer]
Hi @tobbbe, you'll need a `SECURITY DEFINER` function for recursive policies like this. In this case, you can try:
```sql
-- Parameters need to be prefixed because the name clashes with `pm`'s columns
CREATE FUNCTION is_member_of(_user_id uuid, _project_id uuid) RETURNS bool AS $$
SELECT EXISTS (
  SELECT 1
  FROM project_members pm
  WHERE pm.project_id = _project_id
  AND pm.user_id = _user_id
);
$$ LANGUAGE sql SECURITY DEFINER;
```
Then in your `videos` policy definition:
```sql
is_member(auth.uid(), project_id)
```
Likewise, in your `projects` policy definition:
```sql
is_member(auth.uid(), id)
```

See also: https://github.com/supabase/supabase/discussions/3495 https://github.com/supabase/supabase/discussions/3328
